bakery/dbrootkeystore/rootkey.go

```diff
@@ -0,0 +1,360 @@
+}
+
+// NewRootKeys returns a root-keys cache that
+// is limited in size to approximately the given size.
+//
+// The NewStore method returns a store implementation
+// that uses specific store policy and backing database
+// implementation.
+//
+// If clock is non-nil, it will be used to find the current
+// time, otherwise time.Now will be used.
+func NewRootKeys(maxCacheSize int, clock Clock) *RootKeys {
+	if clock == nil {
+		clock = wallClock{}
+	}
+	return &RootKeys{
+		maxCacheSize: maxCacheSize,
+		cache:        make(map[string]RootKey),
+		current:      make(map[Policy]RootKey),
+		clock:        clock,
+	}
+}
+
+// Policy holds a store policy for root keys.
+type Policy struct {
+	// GenerateInterval holds the maximum length of time
+	// for which a root key will be returned from RootKey.
+	// If this is zero, it defaults to ExpiryDuration.
+	GenerateInterval time.Duration
+
+	// ExpiryDuration holds the minimum length of time that
+	// root keys will be valid for after they are returned from
+	// RootKey. The maximum length of time that they
+	// will be valid for is ExpiryDuration + GenerateInterval.
+	ExpiryDuration time.Duration
+}
+
+// NewStore returns a new RootKeyStore implementation that
+// stores and obtains root keys from the given collection.
+//
+// Root keys will be generated and stored following the
+// given store policy.
+//
+// It is expected that all Backing instances passed to a given Store's
+// NewStore method should refer to the same underlying database.
+func (s *RootKeys) NewStore(b Backing, policy Policy) bakery.RootKeyStore {
+	if policy.GenerateInterval == 0 {
+		policy.GenerateInterval = policy.ExpiryDuration
+	}
+	return &store{
+		keys:    s,
+		backing: b,
+		policy:  policy,
+	}
+}
+
+// get gets the root key for the given id, trying the cache first and
+// falling back to calling fallback if it's not found there.
+//
+// If the key does not exist or has expired, it returns
+// bakery.ErrNotFound.
+//
+// Called with s.mu locked.
+func (s *RootKeys) get(id []byte, b Backing) (RootKey, error) {
+	key, cached, err := s.get0(id, b)
+	if err != nil && err != bakery.ErrNotFound {
+		return RootKey{}, errgo.Mask(err)
+	}
+	if err == nil && s.clock.Now().After(key.Expires) {
+		key = RootKey{}
+		err = bakery.ErrNotFound
+	}
+	if !cached {
+		s.addCache(id, key)
+	}
+	return key, err
+}
+
+// get0 is the inner version of RootKeys.get. It returns an item and reports
+// whether it was found in the cache, but doesn't check whether the
+// item has expired or move the returned item to s.cache.
+func (s *RootKeys) get0(id []byte, b Backing) (key RootKey, inCache bool, err error) {
+	if k, ok := s.cache[string(id)]; ok {
+		if !k.IsValid() {
+			return RootKey{}, true, bakery.ErrNotFound
+		}
+		return k, true, nil
+	}
+	if k, ok := s.oldCache[string(id)]; ok {
+		if !k.IsValid() {
+			return RootKey{}, false, bakery.ErrNotFound
+		}
+		return k, false, nil
+	}
+	logger.Infof("cache miss for %q", id)
```

Comment:
I'm not certain we want this logging with this high a priority. Perhaps make it debug.